### PR TITLE
[Merged by Bors] - refactor: generalize Mul of Submodule and SMul of Ideal on Submodule to noncommutative setting

### PR DESCRIPTION
**Pull request by alreadydone**
Instead of using `Submodule.map₂` (which doesn't apply in the noncommutative setting), redefine the `Mul` and `SMul` to be defeq to `AddSubmonoid.mul`. In general, if a semiring A is a module over a semiring R such that `IsScalarTower R A A` is true, then the product of a `Submodule R A` with a `Set A` will be a `Submodule R A`, but we do not need this heterogeneous multiplication yet. We mainly intend to apply this to the case A = R.

---

Preparation for the [Hopkins–Levitzki theorem](https://en.wikipedia.org/wiki/Hopkins%E2%80%93Levitzki_theorem) #17908. E.g. this is necessary to state that the Jacobson radical of an Artinian ring is nilpotent.

- [x] depends on: #18092
- [x] depends on: #18094
- [x] depends on: #18096
- [x] depends on: #18278

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Group/Subgroup/Pointwise.lean [hunk 1]
```diff
@@ -543,4 +543,17 @@ theorem le_pointwise_smul_iff₀ {a : α} (ha : a ≠ 0) {S T : AddSubgroup A} :
 
 end GroupWithZero
 
+section Mul
+
+variable {R : Type*} [NonUnitalNonAssocRing R]
+
+protected def mul : Mul (AddSubgroup R) where
```
Comment (alreadydone) [inline Mathlib/Algebra/Group/Subgroup/Pointwise.lean:550]:
```suggestion
/-- For additive subgroups `S` and `T` of a ring, the product of `S` and `T` as submonoids
is automatically a subgroup, which we define as the product of `S` and `T` as subgroups. -/
protected def mul : Mul (AddSubgroup R) where
```

**Comment by jjaassoonn** [comment]
I might have get the situation completely wrong.
I think now the assumption in "ideal R \bu Submodule R A" is exactly the same as "Set R \bu Submodule R A"in [Submodule.pointwiseSetSMul](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Module/Submodule/Pointwise.html#Submodule.pointwiseSetSMul); do we still need "ideal \bu Submodule R A" at all

**Comment by alreadydone** [comment]
> I might have get the situation completely wrong. I think now the assumption in "ideal R \bu Submodule R A" is exactly the same as "Set R \bu Submodule R A"in [Submodule.pointwiseSetSMul](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Module/Submodule/Pointwise.html#Submodule.pointwiseSetSMul); do we still need "ideal \bu Submodule R A" at all

It's true that "Ideal R \bu Submodule R A" can be made a special case of "Set R \bu Submodule R A" under the current definition of "Set R \bu Submodule R A", but I think maybe "Set R \bu Submodule R A" should only be defined for commutative `R`. Under mathematical convention, `S • M` consists of finite sums of products of the form `s • m` with s in S and m in M. When M is a R-submodule, `S • M` is also a R-submodule if R is commutative, but may not be in general, unless you artificially take the R-span, which is what you do in `Submodule.pointwiseSetSMul`. This makes `S • M` (as a set) dependent on R, and deviates from mathematical convention.

My definition when specialized to the situation "Ideal R \bu Ideal R` is defeq to the multiplication "Submodule R R * Submodule R R" (also generalized to noncommutative setting in this PR). My "Ideal R \bu Submodule R A" factors through "AddSubmonoid R \bu AddSubmonoid A" and "Submodule R A * Submodule R A" factors through "AddSubmonoid A * AddSubmonoid A", so they manifestly don't rely on the R-action. This feature results in nice defeq properties (better than the original `map₂` definition) and allows to prove [`restrictScalars_mul`](https://github.com/leanprover-community/mathlib4/pull/17708/files#diff-b5e8c0cf40e614e95c2884c008cbcaa655890a5cc2722742d81293cfb528bf59R252) etc. by `rfl`.

**Comment by jcommelin** [comment]
Thanks for this PR. I think it is doing good things. But it is also doing a lot.
Could you please split the following parts off, into preparatory PRs?

>     * Change the definition of `1 : Submodule R A` from the range of `algebraMap R A` to the R-span of {1}, which requires fixing many downstream files, but this is the only reasonable definition in the noncommutative setting.
> 
>     * Generalize some lemmas and SMul to AddSubmonoids
> 
>     * Redefine `(Sub)module.annihilator` to work in the noncommutative setting using `RingHom.ker`; move it from Ideal/Operations to Ideal/Maps (which only imports Ideal/Operations) because `RingHom.ker` is defined there.

**Comment by alreadydone** [comment]
> Thanks for this PR. I think it is doing good things. But it is also doing a lot. Could you please split the following parts off, into preparatory PRs?

Good idea; #18092, #18094 and #18096 are created.

**Comment by alreadydone** [comment]
I did further factors and opened dependency PR #18278 (generalize AddSubmonoid Mul lemmas to SMul). The benefit is that the AddSubmonoid SMul lemmas can be applied to both Submodule Mul and Ideal-Submodule SMul. However, due to the lack of a typeclass that provides both left and right distributivity for SMul (unless you go all the way to Module), not all lemmas can be generalized. It would still be possible to unify Submodule Mul and Ideal-Submodule SMul, by generalizing to `SMul (Submodule R A) (Submodule R M)` assuming `Module A M` and `IsScalarTower R A M`. If @eric-wieser and @jcommelin think this is the correct thing to do, I'd oblige.

**Comment by jcommelin** [comment]
> [...] due to the lack of a typeclass that provides both left and right distributivity for SMul (unless you go all the way to Module), not all lemmas can be generalized. It would still be possible to unify Submodule Mul and Ideal-Submodule SMul, by generalizing to `SMul (Submodule R A) (Submodule R M)` assuming `Module A M` and `IsScalarTower R A M`. This will require moving the SMul to the same file as Mul and earlier than Mul.

Please record these observations/todos in the module docstring un `## Implementation notes` and/or `## Todo`.

Otherwise, LGTM

**Comment by alreadydone** [comment]
I've implemented `SMul (Submodule R A) (Submodule R M)` in https://github.com/leanprover-community/mathlib4/commit/e5fd5d60e8c01b2ca87642525c65ab7a1c3e004f, so there's no need of the TODO. I'll add a comment (implementation note?) in the module docstring about AddSubmonoid SMul lemmas.

**Comment by eric-wieser** [comment]
I worry that creates diamonds, so would prefer that commit to be reviewed separately

**Comment by alreadydone** [comment]
> I worry that creates diamonds, so would prefer that commit to be reviewed separately

Okay, opened #18419 for that.

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**